src/lib/OS/PageTable.cc

```diff
@@ -0,0 +1,147 @@
+#include "simeng/OS/PageTable.hh"
+
+#include <cmath>
+#include <iostream>
+#include <vector>
+
+#include "simeng/util/Math.hh"
+
+namespace simeng {
+namespace OS {
+
+bool PageTable::isMapped(uint64_t vaddr) { return find(vaddr) != table_.end(); }
+
+uint64_t PageTable::generateOffsetMask(uint64_t pageSize) {
+  if (!isPow2(PAGE_SIZE)) {
+    std::cerr
+        << "[SimEng:PageTable] Page size should be aligned to a power of 2."
+        << std::endl;
+    std::exit(1);
+  }
+  // The logic below generates a mask with only log2(PAGE_SIZE) number of lowest
+  // significant bits as 1. log2(PAGE_SIZE) gives us the number of bits required
+  // to represent PAGE_SIZE in binary. Once this number is calculated it is used
+  // to perform the logical left shift operation of on ~0, giving us a bit
+  // pattern in which log2(PAGE_SIZE) number of lowest significant bits are 0
+  // and rest are 1. Finally the logical left shifted value is inverted to
+  // retrieve a bit pattern in which log2(PAGE_SIZE) number of lowest
+  // significant bits are 1 and rest are 0.
+  uint64_t lval = std::log2(PAGE_SIZE);
+  uint64_t mask = 0;
+  mask = ~mask;
+  mask = ~(mask << lval);
+  return mask;
```

Comment:
Whilst this logic is correct, couldn't it be simplified to `return (pageSize-1)`? Given we make sure pageSize is a power of 2, we know that pageSize-1 will be a number consisting of all 1s, and will be the correct masking for the number of bits required to represent `pageSize`. i.e. for 4096 = 0b1000000000000; 4095=0b0111111111111

Reply:
done